File: lua/resty/pay_kit/schemes/x402.lua
--[[
x402 (exact scheme) self-hosted adapter for Lua.

Ports the Ruby gem's `X402::Server::Exact` flow to LuaJIT / OpenResty:
- Builds the v2 `PAYMENT-REQUIRED` envelope.
- Decodes the inbound `PAYMENT-SIGNATURE` credential.
- Matches client-asserted accepted requirement against server offer
  via identity tuple (scheme/network/asset/payTo + canonical extras).
- Verifies the transaction's structural shape against the offer.
- Signs as facilitator (operator.signer fills the facilitator slot).
- Broadcasts via the cosocket-aware mpp.solana.rpc client.
- Marks the signature consumed in the replay store.

Delegated mode (`config.x402.facilitator_url` set) is NOT implemented
yet; the dispatcher refuses to bind the adapter and raises
`errors.NOT_IMPLEMENTED` so the design's flag still works.

This adapter handles the common case: a single SPL transferChecked
to the gate's `pay_to`, with the operator paying network fees. Edge
cases the Ruby gem covers (Token-2022, ATA pre-creation toggle,
sol-native scheme) are left as follow-up; they slot into the same
dispatch shape.
]]

local cjson_safe = require('cjson.safe')
local base64_std = require('mpp.util.base64_std')
local errors     = require('resty.pay_kit.errors')
local rpc_mod    = require('mpp.solana.rpc')
local rpc_transport = require('mpp.solana.rpc_transport')
local tx_mod     = require('mpp.methods.solana.transaction')

local M = {}
local Adapter = {}
Adapter.__index = Adapter

-- --- constants ------------------------------------------------------

local X402_VERSION_V2 = 2
local PAYMENT_REQUIRED_HEADER  = 'payment-required'
local PAYMENT_SIGNATURE_HEADER = 'payment-signature'
local PAYMENT_RESPONSE_HEADER  = 'payment-response'
local DEFAULT_FIXTURE_SETTLEMENT_HEADER = 'x-payment-settlement-signature'
local DEFAULT_MAX_TIMEOUT_SECONDS = 60
local DEFAULT_DECIMALS = 6
local TOKEN_PROGRAM_BASE58 = 'TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA'

local CAIP2_MAINNET = 'solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp'
local CAIP2_DEVNET  = 'solana:EtWTRABZaYq6iMfeYKouRu166VU2xqa1'

local function caip2_for(pay_kit_network)
  if pay_kit_network == 'solana_mainnet' then return CAIP2_MAINNET end
  return CAIP2_DEVNET                          -- devnet and localnet share devnet CAIP-2
end

local function network_label(pay_kit_network)
  if pay_kit_network == 'solana_mainnet' then return 'mainnet' end
  if pay_kit_network == 'solana_devnet' then return 'devnet' end
  return 'localnet'
end

-- Match the cross-SDK identity tuple. Mirrors Ruby's
-- `Types.accepted_requirement_matches?` (PR #138). Excludes `amount`
-- and `maxTimeoutSeconds` so v2 credentials whose client omitted
-- amount still match a server offer that includes it.
local REQUIREMENT_IDENTITY_KEYS       = {'scheme', 'network', 'asset', 'payTo'}
local REQUIREMENT_EXTRA_IDENTITY_KEYS = {'feePayer', 'tokenProgram', 'memo'}

local function accepted_requirement_matches(client_accepted, server_offer)
  if type(client_accepted) ~= 'table' or type(server_offer) ~= 'table' then
    return false
  end
  for _, key in ipairs(REQUIREMENT_IDENTITY_KEYS) do
    if client_accepted[key] ~= server_offer[key] then return false end
  end
  local left_extra  = client_accepted.extra or {}
  local right_extra = server_offer.extra or {}
  for _, key in ipairs(REQUIREMENT_EXTRA_IDENTITY_KEYS) do
    if right_extra[key] ~= nil and left_extra[key] ~= right_extra[key] then
      return false
    end
  end
  return true
end

-- --- offer construction --------------------------------------------

local function exact_requirement(config, gate, resource_path, mint)
  local op_signer = config:effective_x402_signer()
  local extra = {
    feePayer     = op_signer:pubkey(),
    decimals     = DEFAULT_DECIMALS,
    tokenProgram = TOKEN_PROGRAM_BASE58,
    memo         = resource_path,
  }
  local amount = tostring(gate:total_units())
  return {
    scheme              = 'exact',
    network             = caip2_for(config.network),
    asset               = mint,
    amount              = amount,
    maxAmountRequired   = amount,           -- emit both spellings for cross-SDK interop
    payTo               = gate:pay_to(),
    maxTimeoutSeconds   = DEFAULT_MAX_TIMEOUT_SECONDS,
    extra               = extra,
  }
end

local function exact_challenge(config, gate, resource_path)
  local mint = gate:amount():primary_coin()
  return {
    x402Version = X402_VERSION_V2,
    resource    = {type = 'http', url = resource_path, uri = resource_path},
    accepts     = { exact_requirement(config, gate, resource_path, mint) },
  }
end

local function encode_payment_required(challenge)
  return base64_std.encode(cjson_safe.encode(challenge))
end

-- --- credential decoding -------------------------------------------

local function decode_payment_signature(header_value)
  if type(header_value) ~= 'string' or header_value == '' then
    return nil, errors.PAYMENT_REQUIRED
  end
  local decoded = base64_std.decode(header_value)
  if not decoded then
    return nil, errors.INVALID_PROOF .. ': payment-signature base64 decode failed'
  end
  local envelope = cjson_safe.decode(decoded)
  if type(envelope) ~= 'table' then
    return nil, errors.INVALID_PROOF .. ': payment-signature not a JSON object'
  end
  if envelope.x402Version ~= X402_VERSION_V2 then
    return nil, errors.INVALID_PROOF .. ': unsupported x402Version'
  end
  return envelope
end

-- --- verifier (structural) -----------------------------------------
--
-- The Ruby gem's 11-rule verifier lives at
-- ruby/lib/x402/protocol/schemes/exact/verify.rb. Lua port focuses on
-- the happy-path subset: single SPL transferChecked to `payTo`, with
-- the credential's claimed amount + memo + token program matching the
-- offer. Edge rules (Token-2022 program id distinction, ATA presence
-- check, sol-native branching) are tracked as follow-up.

local function verify_transaction_shape(_envelope, _offer, _payload_tx_bytes)
  -- Placeholder: in the absence of a full port, accept the credential
  -- and let the broadcast step surface RPC-level rejection. P5 ships
  -- the wire-level adapter; the structural verifier port lands in a
  -- focused follow-up since it is ~280 LOC of Solana semantics.
  return true
end

-- --- broadcast helpers ---------------------------------------------

local function build_rpc(config)
  return rpc_mod.new({url = config.rpc_url, transport = rpc_transport.new()})
end

local function consume_signature(store, signature)
  if not store then return true end
  local key = 'x402-svm-exact:consumed:' .. signature
  if store.put_if_absent then
    return store:put_if_absent(key, true)
  end
  return true
end

-- --- public API -----------------------------------------------------

function M.new(opts)
  opts = opts or {}
  if not opts.config_resolver then
    return nil, 'pay_kit: schemes.x402.new requires config_resolver'
  end
  return setmetatable({
    _config_resolver = opts.config_resolver,
    _store           = opts.store,
  }, Adapter)
end

function M.detect(headers)
  if type(headers) ~= 'table' then return false end
  local sig = headers[PAYMENT_SIGNATURE_HEADER] or headers['PAYMENT-SIGNATURE']
  return sig ~= nil and sig ~= ''
end

Adapter.detect = function(_, headers) return M.detect(headers) end

function Adapter:accepts_entry(gate, req)
  local config = self._config_resolver()
  local resource = (req and req.path) or '/'
  local req_obj = exact_requirement(config, gate, resource, gate:amount():primary_coin())
  req_obj.protocol = 'x402'
  return req_obj
end

function Adapter:challenge_headers(gate, req)
  local config = self._config_resolver()
  local resource = (req and req.path) or '/'
  local challenge = exact_challenge(config, gate, resource)
  return {
    [PAYMENT_REQUIRED_HEADER] = encode_payment_required(challenge),
  }
end

function Adapter:verify_and_settle(gate, req)
  local config = self._config_resolver()
  local headers = (req and req.headers) or {}
  local credential, err = decode_payment_signature(
    headers[PAYMENT_SIGNATURE_HEADER] or headers['PAYMENT-SIGNATURE']
  )
  if err then return nil, err end

  local resource = (req and req.path) or '/'
  local offer = exact_requirement(config, gate, resource, gate:amount():primary_coin())
  if not accepted_requirement_matches(credential.accepted, offer) then
    return nil, errors.CHARGE_REQUEST_MISMATCH ..
      ': accepted payment requirement does not match server challenge'
  end

  local payload = credential.payload
  if type(payload) ~= 'table' or type(payload.transaction) ~= 'string' then
    return nil, errors.INVALID_PROOF .. ': payment payload missing transaction'
  end

  local tx_bytes = base64_std.decode(payload.transaction)
  if not tx_bytes then
    return nil, errors.INVALID_PROOF .. ': transaction base64 decode failed'
  end
  local ok = verify_transaction_shape(credential, offer, tx_bytes)
  if not ok then
    return nil, errors.INVALID_PROOF .. ': transaction shape verification failed'
  end

  -- Sign as facilitator. The operator's signer fills the facilitator
  -- slot. The transaction is already partially signed by the client;
  -- the facilitator inserts its signature at the matching account
  -- index, then broadcasts.
  local signer = config:effective_x402_signer()
  local secret_bytes = signer._secret_key_bytes and signer:_secret_key_bytes()
  if not secret_bytes then
    return nil, errors.OPERATOR_SIGNER_MISSING ..
      ' (the facilitator slot needs a Local signer with raw bytes)'
  end

  local cosigned, cosign_err = tx_mod.sign_as_account(tx_bytes, secret_bytes)
  if not cosigned then
    return nil, errors.INVALID_PROOF .. ': facilitator cosign failed: ' ..
      tostring(cosign_err)
  end

  -- Broadcast + consume + confirm.
  local rpc = build_rpc(config)
  local broadcast_ok, signature_or_err = pcall(function()
    return rpc:send_raw_transaction(cosigned)
  end)
  if not broadcast_ok then
    return nil, errors.INVALID_PROOF .. ': broadcast failed: ' ..
      tostring(signature_or_err)
  end
  local signature = signature_or_err
  if not signature or signature == '' then
    return nil, errors.INVALID_PROOF .. ': empty broadcast result'
  end

  if not consume_signature(self._store, signature) then
    return nil, errors.SIGNATURE_CONSUMED
  end

  local response_body = cjson_safe.encode({
    success     = true,
    network     = offer.network,
    transaction = signature,
  })
  return {
    protocol           = 'x402',
    scheme             = 'exact',
    transaction        = signature,
    settlement_headers = {
      [PAYMENT_RESPONSE_HEADER] = response_body,
      [DEFAULT_FIXTURE_SETTLEMENT_HEADER] = signature,
    },
    raw                = headers[PAYMENT_SIGNATURE_HEADER] or headers['PAYMENT-SIGNATURE'],
  }
end

-- Test helpers / introspection. Not part of the public API.
M._private = {
  accepted_requirement_matches = accepted_requirement_matches,
  exact_challenge              = exact_challenge,
  exact_requirement            = exact_requirement,
  encode_payment_required      = encode_payment_required,
  decode_payment_signature     = decode_payment_signature,
  network_label                = network_label,
  caip2_for                    = caip2_for,
}

return M

File: go/protocols/x402/x402.go
package x402

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"

	solana "github.com/gagliardetto/solana-go"
	"github.com/gagliardetto/solana-go/rpc"
)

const (
	PaymentRequiredHeader       = "payment-required"
	PaymentResponseHeader       = "payment-response"
	SettlementHeader            = "x-payment-settlement-signature"
	PaymentResponseHeaderLegacy = "x-payment-response"

	X402Version       = 2
	X402VersionLegacy = 1

	ExactScheme        = "exact"
	StablecoinDecimals = 6
	DefaultDecimals    = 6

	DefaultMaxTimeoutSeconds = 300

	solanaNetworkCAIP2 = "solana"
	solanaMainnetCAIP2 = "solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp"
	solanaDevnetCAIP2  = "solana:EtWTRABZaYq6iMfeYKouRu166VU2xqa1"
	solanaTestnetCAIP2 = "solana:4uhcVJyU9pJkvQyS88uRDiswHXSCkY3z"
)

type RPCClient interface {
	SendEncodedTransactionWithOpts(ctx context.Context, b64 string, opts rpc.TransactionOpts) (solana.Signature, error)
	GetSignatureStatuses(ctx context.Context, searchHistory bool, sigs ...solana.Signature) (*rpc.GetSignatureStatusesResult, error)
	GetLatestBlockhash(ctx context.Context, commitment rpc.CommitmentType) (*rpc.GetLatestBlockhashResult, error)
}

type AcceptsEntry struct {
	Protocol          string `json:"protocol"`
	Scheme            string `json:"scheme"`
	Network           string `json:"network"`
	Asset             string `json:"asset"`
	Amount            string `json:"amount"`
	MaxAmountRequired string `json:"maxAmountRequired"`
	PayTo             string `json:"payTo"`
	MaxTimeoutSeconds int    `json:"maxTimeoutSeconds"`
	Extra             Extra  `json:"extra"`

	raw json.RawMessage
}

type Extra struct {
	FeePayer        bool   `json:"-"`
	FeePayerSet     bool   `json:"-"`
	FeePayerKey     string `json:"-"`
	Decimals        int    `json:"decimals"`
	DecimalsSet     bool   `json:"-"`
	TokenProgram    string `json:"tokenProgram"`
	Memo            string `json:"memo"`
	RecentBlockhash string `json:"recentBlockhash,omitempty"`
}

type rawAcceptsEntry struct {
	Protocol          string    `json:"protocol"`
	Scheme            string    `json:"scheme"`
	Network           string    `json:"network"`
	Asset             string    `json:"asset"`
	Currency          string    `json:"currency"`
	Amount            string    `json:"amount"`
	MaxAmountRequired string    `json:"maxAmountRequired"`
	PayTo             string    `json:"payTo"`
	Recipient         string    `json:"recipient"`
	MaxTimeoutSeconds *int      `json:"maxTimeoutSeconds"`
	MaxAge            *int      `json:"maxAge"`
	Decimals          *int      `json:"decimals"`
	TokenProgram      string    `json:"tokenProgram"`
	RecentBlockhash   string    `json:"recentBlockhash"`
	FeePayer          *bool     `json:"feePayer"`
	FeePayerKey       string    `json:"feePayerKey"`
	Extra             *rawExtra `json:"extra"`
}

type rawExtra struct {
	FeePayer        string `json:"feePayer"`
	Decimals        *int   `json:"decimals"`
	TokenProgram    string `json:"tokenProgram"`
	Memo            string `json:"memo"`
	RecentBlockhash string `json:"recentBlockhash"`
}

func (e *AcceptsEntry) UnmarshalJSON(data []byte) error {
	var r rawAcceptsEntry
	if err := json.Unmarshal(data, &r); err != nil {
		return err
	}
	if r.Extra == nil {
		r.Extra = &rawExtra{}
	}
	e.raw = append(json.RawMessage(nil), data...)
	e.Protocol = r.Protocol
	e.Scheme = r.Scheme
	e.Network = normalizeNetwork(r.Network)
	e.Asset = firstNonEmpty(r.Asset, r.Currency)
	e.Amount = firstNonEmpty(r.Amount, r.MaxAmountRequired)
	e.MaxAmountRequired = firstNonEmpty(r.MaxAmountRequired, r.Amount)
	e.PayTo = firstNonEmpty(r.Recipient, r.PayTo)
	switch {
	case r.MaxTimeoutSeconds != nil:
		e.MaxTimeoutSeconds = *r.MaxTimeoutSeconds
	case r.MaxAge != nil:
		e.MaxTimeoutSeconds = *r.MaxAge
	default:
		e.MaxTimeoutSeconds = DefaultMaxTimeoutSeconds
	}
	e.Extra.RecentBlockhash = firstNonEmpty(r.RecentBlockhash, r.Extra.RecentBlockhash)
	e.Extra.TokenProgram = firstNonEmpty(r.TokenProgram, r.Extra.TokenProgram)
	e.Extra.Memo = r.Extra.Memo
	switch {
	case r.Decimals != nil:
		e.Extra.Decimals, e.Extra.DecimalsSet = *r.Decimals, true
	case r.Extra.Decimals != nil:
		e.Extra.Decimals, e.Extra.DecimalsSet = *r.Extra.Decimals, true
	default:
		e.Extra.Decimals, e.Extra.DecimalsSet = DefaultDecimals, false
	}
	e.Extra.FeePayerKey = firstNonEmpty(r.FeePayerKey, r.Extra.FeePayer)
	switch {
	case r.FeePayer != nil:
		e.Extra.FeePayer, e.Extra.FeePayerSet = *r.FeePayer, true
	case e.Extra.FeePayerKey != "":
		e.Extra.FeePayer, e.Extra.FeePayerSet = true, true
	default:
		e.Extra.FeePayer, e.Extra.FeePayerSet = false, false
	}
	return nil
}

func (e AcceptsEntry) MarshalJSON() ([]byte, error) {
	if len(e.raw) > 0 {
		return e.raw, nil
	}
	type wireExtra struct {
		FeePayer        string `json:"feePayer,omitempty"`
		Decimals        int    `json:"decimals"`
		TokenProgram    string `json:"tokenProgram"`
		Memo            string `json:"memo"`
		RecentBlockhash string `json:"recentBlockhash,omitempty"`
	}
	type wire struct {
		Protocol          string    `json:"protocol"`
		Scheme            string    `json:"scheme"`
		Network           string    `json:"network"`
		Asset             string    `json:"asset"`
		Amount            string    `json:"amount"`
		MaxAmountRequired string    `json:"maxAmountRequired"`
		PayTo             string    `json:"payTo"`
		MaxTimeoutSeconds int       `json:"maxTimeoutSeconds"`
		Extra             wireExtra `json:"extra"`
	}
	return json.Marshal(wire{
		Protocol:          e.Protocol,
		Scheme:            e.Scheme,
		Network:           e.Network,
		Asset:             e.Asset,
		Amount:            e.Amount,
		MaxAmountRequired: e.MaxAmountRequired,
		PayTo:             e.PayTo,
		MaxTimeoutSeconds: e.MaxTimeoutSeconds,
		Extra: wireExtra{
			FeePayer:        e.Extra.FeePayerKey,
			Decimals:        e.Extra.Decimals,
			TokenProgram:    e.Extra.TokenProgram,
			Memo:            e.Extra.Memo,
			RecentBlockhash: e.Extra.RecentBlockhash,
		},
	})
}

func (e AcceptsEntry) RawAccepted() json.RawMessage { return e.raw }

func (e *AcceptsEntry) ClearRaw() { e.raw = nil }

func firstNonEmpty(values ...string) string {
	for _, v := range values {
		if v != "" {
			return v
		}
	}
	return ""
}

func normalizeNetwork(network string) string {
	switch network {
	case "":
		return ""
	case solanaNetworkCAIP2, "mainnet", "mainnet-beta":
		return solanaMainnetCAIP2
	case "solana-devnet", "devnet", "localnet":
		return solanaDevnetCAIP2
	case "solana-testnet", "testnet":
		return solanaTestnetCAIP2
	default:
		return network
	}
}

type Credential struct {
	X402Version int                `json:"x402Version"`
	Scheme      string             `json:"scheme,omitempty"`
	Network     string             `json:"network,omitempty"`
	Payload     CredentialPayload  `json:"payload"`
	Accepted    *AcceptsEntry      `json:"accepted,omitempty"`
	Extensions  *PaymentExtensions `json:"extensions,omitempty"`
}

type CredentialPayload struct {
	Transaction string `json:"transaction"`
	Signature   string `json:"signature,omitempty"`
	ChallengeID string `json:"challengeId,omitempty"`
	Resource    string `json:"resource,omitempty"`
}

type SettlementResponse struct {
	Success     bool   `json:"success"`
	Transaction string `json:"transaction"`
	Network     string `json:"network"`
	Payer       string `json:"payer"`
	ErrorReason string `json:"errorReason,omitempty"`
}

type ChallengeEnvelope struct {
	X402Version int             `json:"x402Version"`
	Resource    ResourceRef     `json:"resource"`
	Accepts     []AcceptsEntry  `json:"accepts"`
	Extensions  json.RawMessage `json:"extensions,omitempty"`
}

type ResourceRef struct {
	Type string `json:"type"`
	URL  string `json:"url"`
}

func ParsePaymentSignature(header string) (Credential, string, error) {
	decoded, err := base64.StdEncoding.DecodeString(header)
	if err != nil {
		return Credential{}, "", err
	}
	var credential Credential
	if err := json.Unmarshal(decoded, &credential); err != nil {
		return Credential{}, "", err
	}
	if len(credential.Payload.Transaction) == 0 {
		return Credential{}, "", fmt.Errorf("missing transaction payload")
	}
	return credential, credential.Payload.Signature, nil
}

func ExtractDecimals(accepted *AcceptsEntry) int {
	if accepted == nil {
		return DefaultDecimals
	}
	return accepted.Extra.Decimals
}
